overlord/registrystate/registrystate.go

```diff
@@ -197,45 +198,120 @@ var writeDatabag = func(st *state.State, databag registry.JSONDataBag, account,
 	return nil
 }
 
-type cachedRegistryTx struct {
-	account  string
-	registry string
-}
+// GetTransactionToModify retrieves or creates a transaction to change the view's
+// registry. The state must be locked by the caller. Takes a context which should
+// contain a hookstate.Context in it, if it came from a hook context. Calling
+// ctx.Done will trigger the creation of a modify-registry change and wait for
+// its completion (unless one already existed, in which case it just persists
+// changes to the existing transaction).
+func GetTransactionToModify(ctx *Context, st *state.State, view *registry.View) (*Transaction, error) {
+	account, registryName := view.Registry().Account, view.Registry().Name
+
+	// check if we're already running in the context of a committing transaction
+	hookCtx := ctx.hookCtx
+	if IsRegistryHook(hookCtx) {
+		// running in the context of a transaction, so if the referenced registry
+		// doesn't match that tx, we only allow the caller to read the other registry
+		t, _ := hookCtx.Task()
+		tx, commitTask, err := GetStoredTransaction(t)
+		if err != nil {
+			return nil, fmt.Errorf("cannot access registry view %s/%s/%s: cannot get transaction: %v", account, registryName, view.Name, err)
+		}
+
+		if tx.RegistryAccount != account || tx.RegistryName != registryName {
+			return nil, fmt.Errorf("cannot access registry %s/%s: ongoing transaction for %s/%s", account, registryName, tx.RegistryAccount, tx.RegistryName)
+		}
+
+		ctx.onDone(func() error {
+			setTransaction(commitTask, tx)
```

Comment:
this is still a bit strange though as a separation of responsibility, maybe GetStoredTransaction instead of returning commitTask should return a commitModifactions function that does this behind the scenes?

Reply:
Changed and extended a test to cover it as well